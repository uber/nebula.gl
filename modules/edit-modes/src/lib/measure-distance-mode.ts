// @ts-ignore
import turfDistance from '@turf/distance';
// @ts-ignore
import memoize from '../memoize';
import {
  ClickEvent,
  PointerMoveEvent,
  Tooltip,
  ModeProps,
  GuideFeatureCollection,
  EditHandleFeature,
} from '../types';
import { FeatureCollection, Position } from '../geojson-types';
import { GeoJsonEditMode } from './geojson-edit-mode';

const DEFAULT_TOOLTIPS = [];

export class MeasureDistanceMode extends GeoJsonEditMode {
  startingPoint: Readonly<EditHandleFeature> | null | undefined = null;
  endingPoint: Readonly<EditHandleFeature> | null | undefined = null;
  endingPointLocked = false;

  _setEndingPoint(mapCoords: Position) {
    this.endingPoint = {
      type: 'Feature',
      properties: {
        guideType: 'editHandle',
        editHandleType: 'existing',
        featureIndex: -1,
        positionIndexes: [],
      },
      geometry: {
        type: 'Point',
        coordinates: mapCoords,
      },
    };
  }

  _getTooltips = memoize(({ modeConfig, startingPoint, endingPoint }) => {
    let tooltips = DEFAULT_TOOLTIPS;

    if (startingPoint && endingPoint) {
      const { formatTooltip, turfOptions, measurementCallback } = modeConfig || {};
      const units = (turfOptions && turfOptions.units) || 'kilometers';

      const distance = turfDistance(startingPoint, endingPoint, turfOptions);

      let text;
      if (formatTooltip) {
        text = formatTooltip(distance);
      } else {
        // By default, round to 2 decimal places and append units
        text = `${parseFloat(distance).toFixed(2)} ${units}`;
      }

      if (measurementCallback) {
        measurementCallback(distance);
      }

      tooltips = [
        {
          position: endingPoint.geometry.coordinates,
          text,
        },
      ];
    }

    return tooltips;
  });

  handleClick(event: ClickEvent, props: ModeProps<FeatureCollection>): void {
    if (!this.startingPoint || this.endingPointLocked) {
      this.startingPoint = {
        type: 'Feature',
        properties: {
          guideType: 'editHandle',
          editHandleType: 'existing',
          featureIndex: -1,
          positionIndexes: [],
        },
        geometry: {
          type: 'Point',
          coordinates: event.mapCoords,
        },
      };
      this.endingPoint = null;
      this.endingPointLocked = false;
    } else if (this.startingPoint) {
      this._setEndingPoint(event.mapCoords);
      this.endingPointLocked = true;
    }
  }

  // Called when the pointer moved, regardless of whether the pointer is down, up, and whether something was picked
  handlePointerMove(event: PointerMoveEvent, props: ModeProps<FeatureCollection>): void {
    if (this.startingPoint && !this.endingPointLocked) {
      this._setEndingPoint(event.mapCoords);
    }

    props.onUpdateCursor('cell');
  }

  // Return features that can be used as a guide for editing the data
  getGuides(props: ModeProps<FeatureCollection>): GuideFeatureCollection {
    const guides: GuideFeatureCollection = { type: 'FeatureCollection', features: [] };
    const { features } = guides;

    if (this.startingPoint) {
      features.push(this.startingPoint);
    }
    if (this.endingPoint) {
      features.push(this.endingPoint);
    }
    if (this.startingPoint && this.endingPoint) {
      features.push({
        type: 'Feature',
        properties: { guideType: 'tentative' },
        geometry: {
          type: 'LineString',
          coordinates: [
            this.startingPoint.geometry.coordinates,
            this.endingPoint.geometry.coordinates,
          ],
        },
      });
    }
    return guides;
  }

  getTooltips(props: ModeProps<FeatureCollection>): Tooltip[] {
    return this._getTooltips({
      modeConfig: props.modeConfig,
      startingPoint: this.startingPoint,
      endingPoint: this.endingPoint,
    });
  }
}
