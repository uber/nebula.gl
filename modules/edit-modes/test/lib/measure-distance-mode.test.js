// @flow
/* eslint-env jest */

import { MeasureDistanceMode } from '../../src/lib/measure-distance-mode.js';
import {
  createFeatureCollectionProps,
  createClickEvent,
  createPointerMoveEvent,
} from '../test-utils.js';

describe('move without click', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handlePointerMove(createPointerMoveEvent(), createFeatureCollectionProps());
  });

  it('guides are empty', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toEqual({ type: 'FeatureCollection', features: [] });
  });

  it('tooltips are empty', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toEqual([]);
  });
});

describe('one click', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
  });

  it('guides are a single point', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toMatchSnapshot();
  });

  it('tooltips are empty', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toEqual([]);
  });
});

describe('one click + pointer move', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([3, 4]), createFeatureCollectionProps());
  });

  it('guides are two points + line string', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toMatchSnapshot();
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});

describe('two clicks', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
  });

  it('guides are two points + line string', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides).toMatchSnapshot();
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });

  it('can measure kilometers', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips[0].text).toContain('kilometers');
  });

  it('can measure miles', () => {
    const tooltips = mode.getTooltips(
      createFeatureCollectionProps({ modeConfig: { turfOptions: { units: 'miles' } } })
    );
    expect(tooltips[0].text).toContain('miles');
  });

  it('can format distance', () => {
    const tooltips = mode.getTooltips(
      createFeatureCollectionProps({ modeConfig: { formatTooltip: String } })
    );
    expect(tooltips[0].text).toEqual('314.28368918020476');
  });
});

describe('two clicks + pointer move', () => {
  let mode;
  beforeEach(() => {
    mode = new MeasureDistanceMode();
    mode.handleClick(createClickEvent([1, 2]), createFeatureCollectionProps());
    mode.handleClick(createClickEvent([3, 4]), createFeatureCollectionProps());
    mode.handlePointerMove(createPointerMoveEvent([4, 5]), createFeatureCollectionProps());
  });

  it('ending point is clicked point not hovered point', () => {
    const guides = mode.getGuides(createFeatureCollectionProps());
    expect(guides.features[1].geometry.coordinates).toEqual([3, 4]);
  });

  it('tooltip contains distance', () => {
    const tooltips = mode.getTooltips(createFeatureCollectionProps());
    expect(tooltips).toMatchSnapshot();
  });
});
