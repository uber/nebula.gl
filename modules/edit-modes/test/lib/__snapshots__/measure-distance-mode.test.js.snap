// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one click + pointer move guides are two points + line string 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            1,
            2,
          ],
          Array [
            3,
            4,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "guideType": "tentative",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`one click + pointer move tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
]
`;

exports[`one click guides are a single point 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`two clicks + pointer move tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
]
`;

exports[`two clicks guides are two points + line string 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          3,
          4,
        ],
        "type": "Point",
      },
      "properties": Object {
        "editHandleType": "existing",
        "featureIndex": -1,
        "guideType": "editHandle",
        "positionIndexes": Array [],
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            1,
            2,
          ],
          Array [
            3,
            4,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "guideType": "tentative",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`two clicks tooltip contains distance 1`] = `
Array [
  Object {
    "position": Array [
      3,
      4,
    ],
    "text": "314.28 kilometers",
  },
]
`;
